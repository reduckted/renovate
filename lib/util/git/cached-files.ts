import { logger } from '../../logger';
import { getCache } from '../cache/repository';
import type { BranchCache, FileChangeCache } from '../cache/repository/types';
import type { FileChange } from './types';

export function getCachedFiles(branchName: string): FileChangeCache[] {
  const cache = getCache();
  const branch = cache.branches?.find((br) => br.branchName === branchName);
  return branch?.files ? [...branch.files] : [];
}

export function setCachedFiles(branchName: string, files: FileChange[]): void {
  logger.debug('setCachedFiles()');
  const cache = getCache();
  cache.branches ??= [];
  let branch = cache.branches.find((br) => br.branchName === branchName);
  if (!branch) {
    logger.debug(`setCachedFiles(): Branch cache not present`); // should never be called
    branch = {
      branchName,
    } as BranchCache;
    cache.branches.push(branch);
  }

  branch.files = files.map((file) => ({
    type: file.type,
    path: file.path,
  }));
}
