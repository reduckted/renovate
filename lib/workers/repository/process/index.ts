// TODO #22198
import { mergeChildConfig } from '../../../config';
import { configFileNames } from '../../../config/app-strings';
import { GlobalConfig } from '../../../config/global';
import { resolveConfigPresets } from '../../../config/presets';
import type { RenovateConfig } from '../../../config/types';
import { CONFIG_VALIDATION } from '../../../constants/error-messages';
import { addMeta, logger, removeMeta } from '../../../logger';
import type { PackageFile } from '../../../modules/manager/types';
import { platform } from '../../../modules/platform';
import { scm } from '../../../modules/platform/scm';
import { getCache } from '../../../util/cache/repository';
import { clone } from '../../../util/clone';
import { getEnv } from '../../../util/env';
import { getBranchList } from '../../../util/git';
import { addSplit } from '../../../util/split';
import { getRegexPredicate } from '../../../util/string-match';
import { parseConfigsUsingFile } from '../../global/config/parse';
import type { BranchConfig } from '../../types';
import { readDashboardBody } from '../dependency-dashboard';
import { initializeConfig } from '../init';
import { detectVulnerabilityAlerts } from '../init/vulnerability';
import type { ExtractResult } from './extract-update';
import { extract, lookup, update } from './extract-update';
import type { WriteUpdateResult } from './write';

async function getBaseBranchConfig(
  baseBranch: string,
  config: RenovateConfig,
): Promise<RenovateConfig> {
  logger.debug(`baseBranch: ${baseBranch}`);

  let baseBranchConfig: RenovateConfig = clone(config);

  if (
    config.useBaseBranchConfig !== 'none' &&
    baseBranch !== config.defaultBranch
  ) {
    logger.debug(
      { baseBranch },
      config.useBaseBranchConfig === 'merge'
        ? `Merging config from base branch because useBaseBranchConfig=merge`
        : `Replacing config with config from base branch because useBaseBranchConfig=${config.useBaseBranchConfig}`,
    );

    baseBranchConfig = await getConfigFromBaseBranch(baseBranch, config);
    baseBranchConfig = await resolveConfigPresets(baseBranchConfig, config);

    switch (config.useBaseBranchConfig) {
      case 'merge':
        baseBranchConfig = mergeChildConfig(config, baseBranchConfig);
        break;

      case 'replace':
        // Although we are replacing the config with the one from the
        // base branch, we can't use the base branch config as is.
        // We need to apply the defaults, environment and command
        // line arguments to it, in the same way that they were
        // applied to the config from the default branch.
        baseBranchConfig = await parseConfigsUsingFile(
          getEnv(),
          process.argv,
          baseBranchConfig,
        );

        // Once the default configuration has been created,
        // there are some additional changes applied to it.
        baseBranchConfig = initializeConfig(baseBranchConfig);
        baseBranchConfig = await detectVulnerabilityAlerts(baseBranchConfig);
        break;
    }

    // istanbul ignore if
    if (config.printConfig) {
      logger.info(
        { config: baseBranchConfig },
        'Base branch config after ' + config.useBaseBranchConfig,
      );
    }

    // baseBranches value should be based off the default branch
    baseBranchConfig.baseBranches = config.baseBranches;
  }

  if (config.baseBranches!.length > 1) {
    baseBranchConfig.branchPrefix += `${baseBranch}-`;
    baseBranchConfig.hasBaseBranches = true;
  }

  baseBranchConfig = mergeChildConfig(baseBranchConfig, { baseBranch });

  return baseBranchConfig;
}

function unfoldBaseBranches(
  defaultBranch: string,
  baseBranches: string[],
): string[] {
  const unfoldedList: string[] = [];

  const allBranches = getBranchList();
  for (const baseBranch of baseBranches) {
    const isAllowedPred = getRegexPredicate(baseBranch);
    if (isAllowedPred) {
      const matchingBranches = allBranches.filter(isAllowedPred);
      logger.debug(
        `baseBranches regex "${baseBranch}" matches [${matchingBranches.join()}]`,
      );
      unfoldedList.push(...matchingBranches);
    } else if (baseBranch === '$default') {
      logger.debug(`baseBranches "$default" matches "${defaultBranch}"`);
      unfoldedList.push(defaultBranch);
    } else {
      unfoldedList.push(baseBranch);
    }
  }

  return [...new Set(unfoldedList)];
}

async function getConfigFromBaseBranch(
  baseBranch: string,
  config: RenovateConfig,
): Promise<RenovateConfig> {
  // Retrieve config file name autodetected for this repo
  const cache = getCache();
  // TODO: types (#22198)
  const defaultBranchConfigFileName = cache.configFileName!;

  // If we are allowed to detect the config file on the base branch,
  // then we will start by trying each possible config file name.
  let fileNamesToTry: string[] = [];
  if (config.detectBaseBranchConfigFileName) {
    fileNamesToTry.push(...configFileNames);
  }

  // If the config file name from the default branch is not one
  // that we are already going to try, then we'll try it last.
  if (!fileNamesToTry.includes(defaultBranchConfigFileName)) {
    fileNamesToTry.push(defaultBranchConfigFileName);
  }

  // Reading the config from `package.json`
  // is not supported for base branches.
  fileNamesToTry = fileNamesToTry.filter((x) => x !== 'package.json');

  for (const fileName of fileNamesToTry) {
    logger.debug({ baseBranch, fileName }, 'detecting base branch config');
    try {
      const baseBranchConfig = await platform.getJsonFile(
        fileName,
        config.repository,
        baseBranch,
      );
      logger.debug(
        { config: baseBranchConfig, fileName },
        'Base branch config raw',
      );
      return baseBranchConfig;
    } catch {
      // A config file with this name does
      // not exist. Try the next file name.
    }
  }

  logger.error(
    { defaultBranchConfigFileName, baseBranch },
    `Error fetching config file from base branch - possible config name mismatch between branches?`,
  );

  const error = new Error(CONFIG_VALIDATION);
  error.validationSource = 'config';
  error.validationError = 'Error fetching config file';
  error.validationMessage = `Error fetching config file \`${defaultBranchConfigFileName}\` from branch \`${baseBranch}\``;
  throw error;
}

export async function extractDependencies(
  config: RenovateConfig,
  overwriteCache = true,
): Promise<ExtractResult> {
  await readDashboardBody(config);
  let res: ExtractResult = {
    branches: [],
    branchList: [],
    packageFiles: {},
  };
  if (GlobalConfig.get('platform') !== 'local' && config.baseBranches?.length) {
    config.baseBranches = unfoldBaseBranches(
      config.defaultBranch!,
      config.baseBranches,
    );
    logger.debug({ baseBranches: config.baseBranches }, 'baseBranches');
    const extracted: Record<string, Record<string, PackageFile[]>> = {};
    for (const baseBranch of config.baseBranches) {
      addMeta({ baseBranch });
      if (await scm.branchExists(baseBranch)) {
        const baseBranchConfig = await getBaseBranchConfig(baseBranch, config);
        extracted[baseBranch] = await extract(baseBranchConfig, overwriteCache);
      } else {
        logger.warn({ baseBranch }, 'Base branch does not exist - skipping');
      }
    }
    addSplit('extract');
    for (const baseBranch of config.baseBranches) {
      if (await scm.branchExists(baseBranch)) {
        addMeta({ baseBranch });
        const baseBranchConfig = await getBaseBranchConfig(baseBranch, config);
        const packageFiles = extracted[baseBranch];
        const baseBranchRes = await lookup(baseBranchConfig, packageFiles);
        res.branches = res.branches.concat(baseBranchRes?.branches);
        res.branchList = res.branchList.concat(baseBranchRes?.branchList);
        if (!res.packageFiles || !Object.keys(res.packageFiles).length) {
          // Use the first branch
          res.packageFiles = baseBranchRes?.packageFiles;
        }
      }
    }
    removeMeta(['baseBranch']);
  } else {
    logger.debug('No baseBranches');
    const packageFiles = await extract(config, overwriteCache);
    addSplit('extract');
    if (GlobalConfig.get('dryRun') === 'extract') {
      res.packageFiles = packageFiles;
      logger.info({ packageFiles }, 'Extracted dependencies');
      return res;
    }
    res = await lookup(config, packageFiles);
  }
  addSplit('lookup');
  return res;
}

export function updateRepo(
  config: RenovateConfig,
  branches: BranchConfig[],
): Promise<WriteUpdateResult | undefined> {
  logger.debug('processRepo()');

  return update(config, branches);
}
