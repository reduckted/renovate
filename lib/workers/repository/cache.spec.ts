import { REPOSITORY_CHANGED } from '../../constants/error-messages';
import type { Pr } from '../../modules/platform';
import * as _cache from '../../util/cache/repository';
import type {
  BranchUpgradeCache,
  RepoCacheData,
} from '../../util/cache/repository/types';
import * as _behindBaseBranchCache from '../../util/git/behind-base-branch-cache';
import * as _cachedFiles from '../../util/git/cached-files';
import * as _conflictsCache from '../../util/git/conflicts-cache';
import * as _modifiedCache from '../../util/git/modified-cache';
import * as _pristine from '../../util/git/pristine';
import type { LongCommitSha } from '../../util/git/types';
import { setBranchCache } from './cache';
import * as _prCache from './update/pr/pr-cache';
import { logger, partial, platform, scm } from '~test/util';

vi.mock('../../util/cache/repository');
vi.mock('../../util/git/behind-base-branch-cache');
vi.mock('../../util/git/modified-cache');
vi.mock('../../util/git/conflicts-cache');
vi.mock('../../util/git/pristine');
vi.mock('../../util/git/cached-files');
vi.mock('./update/pr/pr-cache');

const cache = vi.mocked(_cache);
const behindBaseBranchCache = vi.mocked(_behindBaseBranchCache);
const modifiedCache = vi.mocked(_modifiedCache);
const conflictsCache = vi.mocked(_conflictsCache);
const pristine = vi.mocked(_pristine);
const cachedFiles = vi.mocked(_cachedFiles);
const prCache = vi.mocked(_prCache);

describe('workers/repository/cache', () => {
  let repoCache: RepoCacheData = {};

  const upgrade: BranchUpgradeCache = {
    datasource: 'a',
    depName: 'b',
    depType: 'c',
    displayPending: 'd',
    packageName: 'e',
    fixedVersion: '2.0.0',
    currentVersion: '1.0.0',
    newVersion: '3.0.0',
    currentValue: 'f',
    newValue: 'g',
    currentDigest: 'h',
    newDigest: 'i',
    packageFile: 'j',
    sourceUrl: 'k',
    remediationNotPossible: true,
    updateType: 'major',
  };

  beforeEach(() => {
    repoCache = {};
    cache.getCache.mockReturnValue(repoCache);

    scm.getBranchCommit.mockImplementation((name) => {
      let sha: string | null = null;
      switch (name) {
        case 'first':
          sha = 'firstSha';
          break;
        case 'second':
          sha = 'secondSha';
          break;
        case 'base':
          sha = 'baseSha';
          break;
      }
      return Promise.resolve(sha as LongCommitSha | null);
    });

    pristine.getCachedPristineResult.mockReturnValue(false);
    platform.getBranchPr.mockResolvedValue(null);
    modifiedCache.getCachedModifiedResult.mockReturnValue(false);
    behindBaseBranchCache.getCachedBehindBaseResult.mockReturnValue(false);
    conflictsCache.getCachedConflictResult.mockReturnValue(false);
    prCache.getPrCache.mockReturnValue(null);
    cachedFiles.getCachedFiles.mockReturnValue([]);
  });

  describe('setBranchCache()', () => {
    it('builds branch cache for the given branches', async () => {
      prCache.getPrCache.mockImplementation((name) => {
        switch (name) {
          case 'first':
            return { bodyFingerprint: 'firstBody', lastEdited: 'firstEdited' };
          default:
            return null;
        }
      });
      cachedFiles.getCachedFiles.mockImplementation((name) => {
        switch (name) {
          case 'first':
            return [
              { path: 'a.txt', type: 'addition' },
              { path: 'b.txt', type: 'deletion' },
            ];
          default:
            return [];
        }
      });

      await setBranchCache([
        {
          branchName: 'first',
          baseBranch: 'base',
          manager: 'test',
          prBlockedBy: 'AwaitingTests',
          prTitle: 'First PR',
          result: 'already-existed',
          prNo: undefined,
          automerge: true,
          upgrades: [
            {
              ...upgrade,
              branchName: 'first',
              manager: 'test',
              displayPending: 'x',
            },
          ],
          commitFingerprint: '123',
        },
        {
          branchName: 'second',
          baseBranch: 'base',
          manager: 'test',
          prBlockedBy: 'AwaitingTests',
          prTitle: 'Second PR',
          result: 'needs-pr-approval',
          prNo: undefined,
          automerge: true,
          upgrades: [
            {
              ...upgrade,
              branchName: 'first',
              manager: 'test',
              displayPending: 'y',
            },
            {
              ...upgrade,
              branchName: 'first',
              manager: 'test',
              displayPending: 'z',
            },
          ],
          commitFingerprint: '123',
        },
      ]);

      expect(repoCache.branches).toMatchSnapshot();
    });

    it('with base branch sha, branch sha, branch pr', async () => {
      platform.getBranchPr.mockResolvedValue(partial<Pr>({ number: 123 }));

      await setBranchCache([
        {
          branchName: 'first',
          baseBranch: 'base',
          manager: 'test',
          prNo: 42,
          upgrades: [],
        },
      ]);

      expect(repoCache.branches).toEqual([
        expect.objectContaining({ prNo: 123 }),
      ]);
    });

    it('with base branch, branch sha, no branch pr', async () => {
      platform.getBranchPr.mockResolvedValue(null);

      await setBranchCache([
        {
          branchName: 'first',
          baseBranch: 'base',
          manager: 'test',
          prNo: 42,
          upgrades: [],
        },
      ]);

      expect(repoCache.branches).toEqual([
        expect.objectContaining({ prNo: null }),
      ]);
    });

    it('with base branch sha, no branch sha, pr number', async () => {
      platform.getBranchPr.mockResolvedValue(partial<Pr>({ number: 123 }));

      await setBranchCache([
        {
          branchName: 'unknown',
          baseBranch: 'base',
          manager: 'test',
          prNo: 42,
          upgrades: [],
        },
      ]);

      expect(repoCache.branches).toEqual([
        expect.objectContaining({ prNo: 42 }),
      ]);
    });

    it('with base branch sha, no branch sha, no pr number', async () => {
      platform.getBranchPr.mockResolvedValue(partial<Pr>({ number: 123 }));

      await setBranchCache([
        {
          branchName: 'unknown',
          baseBranch: 'base',
          manager: 'test',
          prNo: undefined,
          upgrades: [],
        },
      ]);

      expect(repoCache.branches).toEqual([
        expect.objectContaining({ prNo: null }),
      ]);
    });

    it.each([
      { cached: true, expected: true },
      { cached: false, expected: false },
      { cached: null, expected: undefined },
    ])('with isModified=$cached', async ({ cached, expected }) => {
      modifiedCache.getCachedModifiedResult.mockReturnValue(cached);

      await setBranchCache([
        {
          branchName: 'first',
          baseBranch: 'base',
          manager: 'test',
          upgrades: [],
        },
      ]);

      expect(repoCache.branches).toEqual([
        expect.objectContaining({ isModified: expected }),
      ]);
      expect(
        modifiedCache.getCachedModifiedResult,
      ).toHaveBeenCalledExactlyOnceWith('first', 'firstSha');
    });

    it.each([
      { cached: true, expected: true },
      { cached: false, expected: false },
      { cached: null, expected: undefined },
    ])('with isBehindBase=$cached', async ({ cached, expected }) => {
      behindBaseBranchCache.getCachedBehindBaseResult.mockReturnValue(cached);

      await setBranchCache([
        {
          branchName: 'first',
          baseBranch: 'base',
          manager: 'test',
          upgrades: [],
        },
      ]);

      expect(repoCache.branches).toEqual([
        expect.objectContaining({ isBehindBase: expected }),
      ]);
      expect(
        behindBaseBranchCache.getCachedBehindBaseResult,
      ).toHaveBeenCalledExactlyOnceWith('first', 'firstSha', 'base', 'baseSha');
    });

    it.each([
      { cached: true, expected: true },
      { cached: false, expected: false },
      { cached: null, expected: undefined },
    ])('with isConflicted=$cached', async ({ cached, expected }) => {
      conflictsCache.getCachedConflictResult.mockReturnValue(cached);

      await setBranchCache([
        {
          branchName: 'first',
          baseBranch: 'base',
          manager: 'test',
          upgrades: [],
        },
      ]);

      expect(repoCache.branches).toEqual([
        expect.objectContaining({ isConflicted: expected }),
      ]);
      expect(
        conflictsCache.getCachedConflictResult,
      ).toHaveBeenCalledExactlyOnceWith('first', 'firstSha', 'base', 'baseSha');
    });

    it.each([
      { input: true, expected: true },
      { input: false, expected: false },
      { input: undefined, expected: false },
    ])('with automerge=$0', async ({ input, expected }) => {
      await setBranchCache([
        {
          branchName: 'first',
          baseBranch: 'base',
          manager: 'test',
          automerge: input,
          upgrades: [],
        },
      ]);

      expect(repoCache.branches).toEqual([
        expect.objectContaining({ automerge: expected }),
      ]);
    });

    it('throws if repository changed', async () => {
      const error = new Error(REPOSITORY_CHANGED);
      platform.getBranchPr.mockRejectedValue(error);

      await expect(() =>
        setBranchCache([
          {
            branchName: 'first',
            baseBranch: 'base',
            manager: 'test',
            upgrades: [],
          },
        ]),
      ).rejects.toThrowError(error);
    });

    it.each([401, 404])(
      'logs warning and does not cache branch if HTTP $0 error occurs',
      async (statusCode) => {
        const error = Object.assign(new Error(), { response: { statusCode } });
        platform.getBranchPr.mockRejectedValue(error);

        await setBranchCache([
          {
            branchName: 'first',
            baseBranch: 'base',
            manager: 'test',
            upgrades: [],
          },
        ]);

        expect(repoCache.branches).toEqual([]);
        expect(logger.logger.warn).toHaveBeenCalledWith(
          { err: error, branchName: 'first' },
          'HTTP error generating branch cache',
        );
      },
    );

    it('logs error and does not cache branch if other error occurs', async () => {
      const error = Object.assign(new Error(), { statusCode: 500 });
      platform.getBranchPr.mockRejectedValue(error);

      await setBranchCache([
        {
          branchName: 'first',
          baseBranch: 'base',
          manager: 'test',
          upgrades: [],
        },
      ]);

      expect(repoCache.branches).toEqual([]);
      expect(logger.logger.error).toHaveBeenCalledWith(
        { err: error, branchName: 'first' },
        'Error generating branch cache',
      );
    });
  });
});
