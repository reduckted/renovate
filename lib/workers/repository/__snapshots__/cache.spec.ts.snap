// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`workers/repository/cache > setBranchCache() > builds branch cache for the given branches 1`] = `
[
  {
    "automerge": true,
    "baseBranch": "base",
    "baseBranchSha": "baseSha",
    "branchName": "first",
    "commitFingerprint": "123",
    "files": [
      {
        "path": "a.txt",
        "type": "addition",
      },
      {
        "path": "b.txt",
        "type": "deletion",
      },
    ],
    "isBehindBase": false,
    "isConflicted": false,
    "isModified": false,
    "prBlockedBy": "AwaitingTests",
    "prCache": {
      "bodyFingerprint": "firstBody",
      "lastEdited": "firstEdited",
    },
    "prNo": null,
    "prTitle": "First PR",
    "pristine": false,
    "result": "already-existed",
    "sha": "firstSha",
    "upgrades": [
      {
        "currentDigest": "h",
        "currentValue": "f",
        "currentVersion": "1.0.0",
        "datasource": "a",
        "depName": "b",
        "depType": "c",
        "displayPending": "x",
        "fixedVersion": "2.0.0",
        "newDigest": "i",
        "newValue": "g",
        "newVersion": "3.0.0",
        "packageFile": "j",
        "packageName": "e",
        "remediationNotPossible": true,
        "sourceUrl": "k",
        "updateType": "major",
      },
    ],
  },
  {
    "automerge": true,
    "baseBranch": "base",
    "baseBranchSha": "baseSha",
    "branchName": "second",
    "commitFingerprint": "123",
    "files": [],
    "isBehindBase": false,
    "isConflicted": false,
    "isModified": false,
    "prBlockedBy": "AwaitingTests",
    "prCache": null,
    "prNo": null,
    "prTitle": "Second PR",
    "pristine": false,
    "result": "needs-pr-approval",
    "sha": "secondSha",
    "upgrades": [
      {
        "currentDigest": "h",
        "currentValue": "f",
        "currentVersion": "1.0.0",
        "datasource": "a",
        "depName": "b",
        "depType": "c",
        "displayPending": "y",
        "fixedVersion": "2.0.0",
        "newDigest": "i",
        "newValue": "g",
        "newVersion": "3.0.0",
        "packageFile": "j",
        "packageName": "e",
        "remediationNotPossible": true,
        "sourceUrl": "k",
        "updateType": "major",
      },
      {
        "currentDigest": "h",
        "currentValue": "f",
        "currentVersion": "1.0.0",
        "datasource": "a",
        "depName": "b",
        "depType": "c",
        "displayPending": "z",
        "fixedVersion": "2.0.0",
        "newDigest": "i",
        "newValue": "g",
        "newVersion": "3.0.0",
        "packageFile": "j",
        "packageName": "e",
        "remediationNotPossible": true,
        "sourceUrl": "k",
        "updateType": "major",
      },
    ],
  },
]
`;
